File: flutter_app/lib/shared/models/pfz_zone.dart
class PfzZone {
  final String id;
  final String state;
  final double confidence;
  final String source;
  final List<List<double>> polygon;
  final double centerLat;
  final double centerLon;
  final Map<String, double> speciesProbability;
  final double sst;
  final double chlorophyll;
  final DateTime validFrom;
  final DateTime validUntil;
  final double distanceFromShoreKm;
  final String safetyStatus;
  final double? distanceFromUserKm;

  const PfzZone({
    required this.id,
    required this.state,
    required this.confidence,
    required this.source,
    required this.polygon,
    required this.centerLat,
    required this.centerLon,
    required this.speciesProbability,
    required this.sst,
    required this.chlorophyll,
    required this.validFrom,
    required this.validUntil,
    required this.distanceFromShoreKm,
    required this.safetyStatus,
    this.distanceFromUserKm,
  });

  factory PfzZone.fromJson(Map<String, dynamic> json) {
    final environmentalFactors =
        (json['environmental_factors'] as Map?)?.cast<String, dynamic>() ?? {};
    final center = (json['center'] as List?) ?? const [0.0, 0.0];
    return PfzZone(
      id: json['zone_id'] as String,
      state: json['state'] as String,
      confidence: (json['confidence'] as num).toDouble(),
      source: (json['source'] as String?) ?? 'unknown',
      polygon: ((json['polygon'] as List?) ?? const [])
          .map((p) => (p as List).map((v) => (v as num).toDouble()).toList())
          .toList(),
      centerLat: (center[1] as num).toDouble(),
      centerLon: (center[0] as num).toDouble(),
      speciesProbability: Map<String, double>.from(
        ((json['species_probability'] as Map?) ?? const {}).map(
          (k, v) => MapEntry(k as String, (v as num).toDouble()),
        ),
      ),
      sst: ((environmentalFactors['sst'] as num?) ?? 0).toDouble(),
      chlorophyll: ((environmentalFactors['chlorophyll'] as num?) ?? 0).toDouble(),
      validFrom: DateTime.parse(json['valid_from'] as String),
      validUntil: DateTime.parse(json['valid_until'] as String),
      distanceFromShoreKm: ((json['distance_from_shore_km'] as num?) ?? 0).toDouble(),
      safetyStatus: (json['safety_status'] as String?) ?? 'unknown',
      distanceFromUserKm: json['distance_from_user_km'] != null
          ? (json['distance_from_user_km'] as num).toDouble()
          : null,
    );
  }

  String get topSpecies {
    if (speciesProbability.isEmpty) return '';
    return speciesProbability.entries
        .reduce((a, b) => a.value > b.value ? a : b)
        .key;
  }

  bool get isValid => DateTime.now().isBefore(validUntil);
}

File: backend/app/models/pfz_zone.py
from sqlalchemy import Column, String, Float, DateTime, JSON, Index, text
from sqlalchemy.dialects.postgresql import UUID
from geoalchemy2 import Geometry
from datetime import datetime
import uuid

from ..core.database import Base


class PfzZone(Base):
    __tablename__ = "pfz_zones"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    zone_id = Column(String, unique=True, nullable=False, index=True)
    state = Column(String(50), nullable=False, index=True)
    confidence = Column(Float, nullable=False)
    source = Column(String(100), default="ml_model_v3.0")
    polygon = Column(Geometry("POLYGON", srid=4326), nullable=False)
    center = Column(Geometry("POINT", srid=4326), nullable=False)
    species_probability = Column(JSON, default={})
    environmental_factors = Column(JSON, default={})
    valid_from = Column(DateTime(timezone=True), nullable=False)
    valid_until = Column(DateTime(timezone=True), nullable=False)
    distance_from_shore_km = Column(Float)
    recommended_boat_type = Column(String(50))
    safety_status = Column(String(20), default="green")
    created_at = Column(DateTime, default=datetime.utcnow)

    __table_args__ = (
        Index("idx_pfz_state_valid", "state", "valid_until"),
        Index("idx_pfz_geom", "polygon", postgresql_using="gist"),
        Index("idx_pfz_center", "center", postgresql_using="gist"),
        {"timescaledb_hypertable": False},  # TimescaleDB note
    )


class OceanData(Base):
    __tablename__ = "ocean_data"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    timestamp = Column(DateTime(timezone=True), nullable=False, index=True)
    state = Column(String(50), index=True)
    location = Column(Geometry("POINT", srid=4326))
    sst = Column(Float)           # Sea Surface Temperature (°C)
    chlorophyll = Column(Float)   # mg/m³
    current_u = Column(Float)     # m/s eastward
    current_v = Column(Float)     # m/s northward
    wave_height = Column(Float)   # meters
    wind_speed = Column(Float)    # km/h
    wind_direction = Column(Float) # degrees
    source = Column(String(100))

    __table_args__ = (
        Index("idx_ocean_time_state", "timestamp", "state"),
        Index("idx_ocean_geom", "location", postgresql_using="gist"),
    )
